src/core/src/pass/constant_folding.cpp

```diff
@@ -105,7 +105,8 @@ bool ov::pass::ConstantFolding::run_on_model(const std::shared_ptr<ov::Model>& m
 
     for (const auto& original_node : model->get_ordered_ops()) {
         auto node = original_node;
-        if (node_has_requires_precision_conversion_attribute(node)) {
+        if (original_node->can_constant_fold(original_node->input_values()) &&
+            node_has_requires_precision_conversion_attribute(node)) {
```

Comment:
I propose to refactor the order as well, otherwise there is a kind of mess of the performed actions.
What about the following order:
1) If the node cannot be folded, then go recursive into `MultiSubGraphOp` and then `restore_original_input_precision` and `continue`
2) `convert_to_supported_precision` if necessery
3) constant fold the node

Reply:
@itikhono do you agree with the proposed sequence of actions?

Reply:
Absolutely, it looks like, now we might skip MultiSubGraphOp ops and do not run ConstantFolding for graph inside

Reply:
Can we really skip it? From my understanding if a node is actually MultiSubGraphOp, then, if the node itself cannot be folded, we still need to go inside and fold possible constant subgraphs within MultiSubGraphOp

Reply:
I mean we might skip the graph inside of MultiSubGraphOp

Reply:
@jane-intel @praasz the refactoring suggested above requires `can_constant_fold` overriding for all nodes `constant_fold` is overrided for.
I updated the PR with the example of `can_constant_fold` overriding for `ShapeOf`. The same is required for ConvertLike, Reshape, Squeeze, StridedSlice nodes.
Are such changes acceptable in such PR?